lib/hanami.rb

```diff
@@ -164,7 +164,7 @@ def self.env?(*names)
   #     end
   #   end
   #
-  # @return [Hanami::Logger]
+  # @return [Dry::Logger::Dispatcher]
```

Comment:
Question about the multi-backend dispatching that we have in place with `Dry::Logger::Dispatcher`: do we actually need it? It does make dry-logger more challenging to understand (think about this case right here, a user may ask "what is a  _dispatcher_ and why am I getting one of them instead of an ordinary _logger?_") for what feels like a fairly obtuse feature. What did you have in mind when adding this in, @solnic?

Reply:
@solnic I agree.

Reply:
Shouldn't `Dry::Logger` and `Dry::Logger::Dispatcher` implement the same interface? So that we have `Dry::Logger` in annotations but the end user can swap a simple logger with one having multiple backends (with whatever DSL we provide them with)? The notion of multiple backends shouldn't affect the app in any way, right? @solnic @timriley

Reply:
for the record, I haven't seen dry-logger yet :)

Reply:
The whole purpose of dry-logger is to provide a multi-backend logging, otherwise you can just use logger from stdlib. This is extremely useful as it allows you to configure different types of logging for different environments, contexts, sub-apps etc.

This gives you very powerful logging capabilities and I **really** want Hanami to have this provided by default as it will be a big advantage.

Logging/observability has been typically very primitive in the Rails world precisely due to having just one logger that has one destination. Then people would struggle adding other loggers or even other gems that provide more powerful logging features (ie see https://logger.rocketjob.io/). This was incoherent and because of this I've a strong opinion that a framework should provide this built-in, with conventions how things should be configured.

I can give you a simple example that's pretty common these days where advanced logging is needed: you run things on GCP, it's got very powerful logging/alerts/metrics features but you have to log in a format that it understands. It's *not needed* in local environments though. So, how do you do this? If you have only one logger with one destination then you need a different logger for production env. This is really cumbersome to handle though and it's easy to end up with inconsistent logging. Now, when you have a powerful logger with support for multiple destinations, you simply set up a GCP destination (aka backend) in production and you're done. Another use case is exception logger - on production you can add a destination logger that logs exceptions to ie Rollbar or Sentry *but* in development you see exceptions in a simple log file. This kind of scenarios are much simpler to handle when you have a unified logging interface!

This is a real game-changer when it comes to logging and it will be, for sure, one of the selling points of Hanami.

Reply:
@solnic I didn't question the idea of multiple backends, I can provide you with my examples if you will :) The question was how a dispatcher's interface differs from an ordinary logger for the end user.

There are two stages of working with a logger:
1. Making a logger. This may involve an arbitrary API for setting multiple backends or what have you. It returns a, say, `Dry::Logger`.
2. Working with a logger, that is calling `.info`, `debug`, and so on. In other words, a standard logger interface.

It seems like we conflate the two here.
I think in most the end user should see the second version, which is `Dry::Logger`.
Am I making sense?

Reply:
`Dry::Logger` is a namespace, it's not a class. `Dry::Logger.setup` should be the primary interface for setting up a logger. The only reason why there's `Dry::Logger.new` is because originally I've made it like that to maintain compatibility with `Hanami::Logger`. Now it should be just removed. The `Dry::Logger::Dispatcher` implements the same interface as every logging backend, so `info`, `debug` etc. are available and it simply forwards to configured backends.

Reply:
Thanks for the explanation, @solnic :) Definitely sounds like you've thought about this a lot more than me, so I'm happy to trust your instincts here.

I guess this just leaves the matter of the `Dry::Logger::Dispatcher` name. In isolation I don't mind that name at all — it describes its role well. So I think if we're going to advertise the Hanami logger as possessing additional helpful capabilities like multiple backends, then we may as well just lean into this and expose the `Dispatcher` name wherever it makes sense.

I'm comfortable with this now.

Reply:
@timriley we *could* just do `Hanami::Logger < Dry::Logger::Dispatcher` but I dunno

Reply:
@solnic That feels like it'd do more harm than good in the long run, I'm fine with Dry::Logger::Dispatcher